src/shared/boot-entry.c

```diff
@@ -0,0 +1,220 @@
+                        return r;
+
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "No machine ID set, and %s/etc/os-release carries no ID=/IMAGE_ID= fields.",
+                                       strempty(root));
+
+        case BOOT_ENTRY_TOKEN_MACHINE_ID:
+                r = entry_token_from_machine_id(machine_id, type, ret);
+                if (r != 0)
+                        return r;
+
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "No machine ID set.");
+
+        case BOOT_ENTRY_TOKEN_OS_IMAGE_ID:
+                r = entry_token_from_os_release(root, type, ret);
+                if (r != 0)
+                        return r;
+
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "IMAGE_ID= field not set in %s/etc/os-release.",
+                                       strempty(root));
+
+        case BOOT_ENTRY_TOKEN_OS_ID:
+                r = entry_token_from_os_release(root, type, ret);
+                if (r != 0)
+                        return r;
+
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "ID= field not set in %s/etc/os-release.",
+                                       strempty(root));
+
+        case BOOT_ENTRY_TOKEN_LITERAL:
+                return 0;
+
+        default:
+                assert_not_reached();
+        }
+}
+
+int parse_boot_entry_token_type(const char *s, BootEntryTokenType *type, char **token) {
+        assert(s);
+        assert(type);
+        assert(token);
+
+        /*
+         * This function is intended to be used in command line parsers, to handle token that are passed in.
+         *
+         * NOTE THAT THIS WILL FREE THE PREVIOUS ARGUMENT POINTER ON SUCCESS!
+         * Hence, do not pass in uninitialized pointers.
+         */
+
+        if (streq(s, "machine-id")) {
+                *type = BOOT_ENTRY_TOKEN_MACHINE_ID;
+                *token = mfree(*token);
+                return 0;
+        }
+
+        if (streq(s, "os-image-id")) {
+                *type = BOOT_ENTRY_TOKEN_OS_IMAGE_ID;
+                *token = mfree(*token);
+                return 0;
+        }
+
+        if (streq(s, "os-id")) {
+                *type = BOOT_ENTRY_TOKEN_OS_ID;
+                *token = mfree(*token);
+                return 0;
+        }
+
+        const char *e = startswith(s, "literal:");
+        if (e) {
+                if (!boot_entry_token_valid(e))
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                               "Invalid entry token literal is specified for --entry-token=.");
+
+                *type = BOOT_ENTRY_TOKEN_LITERAL;
+                return free_and_strdup_warn(token, e);
+        }
+
+        return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                               "Unexpected parameter for --entry-token=: %s", s);
```

Comment:
Hmm, this seems iffy to have this message in what is now a library function. Shouldn't this just return EINVAL and the let the caller log on its own? Similarly, it is against our usual style to do `mfree` on the output arg in a library function. Shouldn't the caller use temporary variable and handle the replacement on its own?

Reply:
The function is based on `parse_path_argument()` or similar in parse-argument.c. And, the 'unusual' behavior is already commented at the beginning of the function.

Reply:
I see now, thanks.